src/pkg/upgrading/upgrader.go

```diff
@@ -0,0 +1,92 @@
+package upgrading
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/defenseunicorns/go-oscal/src/internal/utils"
+	"github.com/defenseunicorns/go-oscal/src/pkg/validation"
+	"gopkg.in/yaml.v3"
+)
+
+type Upgrader struct {
+	bytes []byte
+	validation.Validator
+	modelVersion    string
+	upgradedJsonMap map[string]interface{}
+}
+
+// NewUpgrader returns an upgrader with a validator created from the desired version.
+func NewUpgrader(model utils.InterfaceOrBytes, desiredVersion string) (upgrader Upgrader, err error) {
+	validator, err := validation.NewValidatorDesiredVersion(model, desiredVersion)
+	if err != nil {
+		return upgrader, err
+	}
+
+	version, err := utils.GetOscalVersionFromMap(validator.GetJsonModel())
+	if err != nil {
+		return upgrader, err
+	}
+
+	return Upgrader{
+		Validator:       validator,
+		modelVersion:    version,
+		upgradedJsonMap: nil,
+	}, nil
+}
+
+// GetModelVersion returns the version of the model being upgraded.
+func (u *Upgrader) GetModelVersion() string {
+	return u.modelVersion
+}
+
+// GetUpgradedJsonMap sets the upgraded model.
+func (u *Upgrader) GetUpgradedJsonMap() map[string]interface{} {
+	return u.upgradedJsonMap
+}
+
+func (u *Upgrader) Upgrade() (err error) {
```

Comment:
What happens when we need to support an upgrade across breaking changes? How do we identify that certain circumstances apply?

if a field was removed - we'd want to add warnings to signify. What if a field is migrated? could we perform the migration of the data from one location to another?

A lot of what-ifs that don't need to be solved now - just curious if we have thoughts while the oven is warm

Reply:
Right now that is not possible, I think there would need to be a lot of work invested into upgrading or handling these changes after a release is put out, right now this will work regardless by providing the things that need to change when revising to a new version, then handling the revision once the validation is successful. I agree that we need better guidance for the result. Mostly it would just come down to maintainer effort and response to upgraded versions, but there is still a lot we likely could not automate without moving into an interactive mode and walking through the errors so the user could provide direct input. (long term we may be able to leverage LAI at least as its concerned with uds, to help with field generation etc)

Reply:
My current thoughts would be breaking into an interactive mode that consumes the result and attempts to allow the user to fill the fields/move them around if it makes sense. 

I also think #92 could be a really good step forward in ensuring the user has clear step by step actions to take if a revision fails.

Reply:
I would like to bring a UX designer in on that story, to work with some ideas of how the user could get and utilize the response clearly and effectively.

Reply:
Because they are essentially just linting errors, and so far even co-pilot, vs-code and other tools don't really do much more that tell me where and what is broken.